### PR TITLE
Unified cache maintenance services for OP-TEE core

### PR DESCRIPTION
**Pull request by etienne-lms**
Provides cache maintenance service to OP-TEE core. In current OP-TEE core, unified cache maintenance services are only provided to TAs through a SVC syscall. This PR aims at providing the feature to both OP-TEE core and OP-TEE TA.

This PR is extracted from https://github.com/OP-TEE/optee_os/pull/1322 and supersedes it.

### CODE REVIEW

**Comment by etienne-lms** [comment]
This change is required by the SDP feature, to allow pseudo TAs to flush/clean/invalidate SDP memory when a SDP pTA is invoked to access (read and/or write) a SDP buffer.

Actually, these changes are useful only to the platforms relying on a outer cache.

File: core/arch/arm/tee/cache.c [hunk 1]
```diff
@@ -49,13 +49,15 @@ TEE_Result cache_operation(enum utee_cache_operation op, void *va, size_t len)
 
 	switch (op) {
 	case TEE_CACHEFLUSH:
+#ifdef CFG_PL310
```
Comment (jforissier) [inline core/arch/arm/tee/cache.c]: I think this is correct, but I am concerned with the terminology. I mean, `cache_maintenance_l2()` does nothing when `CFG_PL310` is not set, yet the platform is very likely to have a L2 cache. I find it confusing. I think we should introduce "outer" in the function name. How about `cache_maintenance_outer_l2()`?

Comment (etienne-lms) [inline core/arch/arm/tee/cache.c]: Yes, I agree. This is indeed confusing. Your proposal about renaming `cache_maintenance_l2()` into `cache_maintenance_outer_l2()` is nice. i will apply it.

Comment (jforissier) [inline core/arch/arm/tee/cache.c]: OK, good. Please make the rename a separate patch, before this one (or after, as you prefer).

Comment (etienne-lms) [inline core/arch/arm/tee/cache.c]: By the way, we currently do not define a `CFG_OUTER_CACHE` configuration directive, as only PL310 is supported as outer cache and `CFG_PL310` cover the case. I prefer to keep relying on `CFG_PL310,` rather than introducing an extra (and almost useless) `CFG_OUTER_CACHE.` Do you agree ?

Comment (jforissier) [inline core/arch/arm/tee/cache.c]: Yes, let's just keep `CFG_PL310`.

**Comment by jforissier** [comment]
Looks good to me.
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
The outer cache could just as well be a level 3 cache (dependent on the number of inner caches), which level the outer cache represent isn't valid here.

While we're at it could I suggest aligning the a32 implementation with the a64 implementation, at least featurewise (deal with all levels of the inner cache, not assuming there's only one level)? I suppose the v7 implementation used in (or prepared for) ARM-TF would be nice to import.

**Comment by jenswi-linaro** [comment]
Switching implementation for v7 cache maintenance could be done in another PR.

**Comment by jforissier** [comment]
@jenswi-linaro so for the last commit in particular, are you recommending `s/cache_maintenance_outer_l2/cache_maintenance_outer/`? I'm OK with such a change, and as you said it could be a first step before `s/cache_maintenance_l1/cache_maintenance_inner/` (if I understand correctly).

**Comment by etienne-lms** [comment]
I think current `cache_maintenance_l1()` acts on level1 cache, not level2 even if it is inner cache.

**Comment by etienne-lms** [comment]
I corsscheck and I feel i was wrong. `cache_maintenance_l1()` acts on data cache, all levels of the unified cache.

**Comment by jforissier** [comment]
@etienne-lms : @jenswi-linaro is *never* wrong ;)

**Comment by jenswi-linaro** [comment]
@jforissier yes about the outer cache, with the flattery you're exaggerating. :-)
 
@etienne-lms good, then the a32 inner cache maintenance implementation is OK.

**Comment by etienne-lms** [comment]
Je l'aurai, un jour. je l'aurai !  *  


(* french stupid commercial joke)

**Comment by etienne-lms** [comment]
As i am renaming `cache_maintenance_l2()`, i can straight move to `cache_maintenance_inner/outer()`, as jerome suggested.

File: core/arch/arm/mm/core_mmu.c [hunk 2]
```diff
@@ -648,7 +648,7 @@ int core_tlb_maintenance(int op, unsigned int a)
 	return 0;
 }
 
-unsigned int cache_maintenance_l1(int op, void *va, size_t len)
+unsigned int cache_maintenance_l1(enum cache_op op, void *va, size_t len)
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu.c]: Please change the return type also.

File: core/arch/arm/mm/core_mmu.c [hunk 3]
```diff
@@ -690,7 +690,8 @@ unsigned int cache_maintenance_l1(int op, void *va, size_t len)
 }
 
 #ifdef CFG_PL310
-unsigned int cache_maintenance_l2(int op, paddr_t pa, size_t len)
+unsigned int cache_maintenance_outer_l2(enum cache_op op,
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu.c]: Please change the return type also.

Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: yes, thanks.

File: core/arch/arm/mm/core_mmu.c (hunk 2 above)
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c]: right!

**Comment by jforissier** [comment]
There is a typo in the commit message of your last update (`s/cache_op_outer/cache_op_inner/`).
Other than that, my R-b still applies.

**Comment by etienne-lms** [comment]
Rebased (last commit is split in 3), tags applied (jens, I added your tag, I hope i did not wrong)
Travis was only complaining on a URL exceeding 80 chars (not fixed) and on space/tab (fixed).

Tested on b2260 (through the SDP test setup I must admit).

**Comment by etienne-lms** [comment]
@jforissier : I fixed the typo in the commit message you pointed out.

**Comment by etienne-lms** [comment]
(updated with jerome sole tag)

File: core/include/tee/svc_cache.h [hunk 4]
```diff
@@ -31,7 +31,8 @@
 #include <tee_api_types.h>
 
 #ifdef CFG_CACHE_API
-TEE_Result syscall_cache_operation(void *va, size_t len, unsigned long op);
+TEE_Result syscall_cache_operation(void *va, size_t len,
+				   enum utee_cache_operation op);
```
Comment (jenswi-linaro) [inline core/include/tee/svc_cache.h]: I'd rather keep this as `unsigned long` instead of `enum utee_cache_operation` for two reasons, it's still `unsigned long` in `lib/libutee/include/utee_syscalls.h` and then there's also the comment at the beginning of that .h file.

File: core/arch/arm/include/mm/core_mmu.h [hunk 5]
```diff
@@ -375,23 +361,36 @@ enum teecore_tlb_op {
...
+enum cache_op {
+	DCACHE_CLEAN,
+	DCACHE_AREA_CLEAN,
+	DCACHE_INVALIDATE,
+	DCACHE_AREA_INVALIDATE,
+	ICACHE_INVALIDATE,
+	ICACHE_AREA_INVALIDATE,
+	DCACHE_CLEAN_INV,
+	DCACHE_AREA_CLEAN_INV,
+	L2CACHE_INVALIDATE,
```
Comment (jenswi-linaro) [inline core/arch/arm/include/mm/core_mmu.h]: Do we need these `L2CACHE_` enums? The outer cache is a data cache.

File: core/include/tee/svc_cache.h (hunk 4 above)
Comment (etienne-lms) [inline core/include/tee/svc_cache.h]: I see your point. Ok to restore `unsigned long op`.

File: core/arch/arm/include/mm/core_mmu.h (hunk 5 above)
Comment (etienne-lms) [inline core/arch/arm/include/mm/core_mmu.h]: No, we don't. That's only legacy. Ok to clean it and move to `DCACHE_xxx` for both inner and outer cache data maintenance operations.

File: core/arch/arm/tee/svc_cache.c [hunk 6]
```diff
@@ -60,5 +59,5 @@ TEE_Result syscall_cache_operation(void *va, size_t len,
 	if (res != TEE_SUCCESS)
 		return TEE_ERROR_ACCESS_DENIED;
 
-	return cache_operation(op, va, len);
+	return cache_operation((enum utee_cache_operation)op, va, len);
```
Comment (jenswi-linaro) [inline core/arch/arm/tee/svc_cache.c]: Is this cast needed?

Comment (etienne-lms) [inline core/arch/arm/tee/svc_cache.c]: strictly speaking, `cache_operation()` expects an ID from enumerated `utee_cache_operation.` Hence i would rather cast here so that the call fulfils the function prototype. I considered  `cache_operation()` as a core internal routine while `syscall_cache_operation()` is the syscall handler, hence conforming the `unsigned long` type used in the ABI.

Comment (jenswi-linaro) [inline core/arch/arm/tee/svc_cache.c]: So what is the purpose of this cast? Does it have any technical advantage or is it purely cosmetic?

Comment (etienne-lms) [inline core/arch/arm/tee/svc_cache.c]: It is cosmetic. Could it help compiler in some way ? Keeping or removing this cast is a matter of taste ? (i feel confortable with either cases)

Comment (jenswi-linaro) [inline core/arch/arm/tee/svc_cache.c]: Unnecessary casting could hide some warning and also clutters the code in my opinion. I doubt you'll find much of this pattern in the Linux kernel.

Comment (etienne-lms) [inline core/arch/arm/tee/svc_cache.c]: ok, then i discard this change. thanks.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
@etienne-lms my R-b's still apply.

**Comment by etienne-lms** [comment]
rebased, tags applied.

**Comment by etienne-lms** [comment]
expecting travis to fail only on comment where a URL exceeds 80 chars.
